docs - clarify caching policy interactions (#45584)
metabase/metabase: docs/configuring-metabase/caching.md

If your question results don't change frequently, you may want to store the results so that the next time anyone visits the question, Metabase can retrieve the cached results rather than query the database again.

For example, if your data only updates once a day, there's no point in querying the database more than once a day, as the data won't have changed. Returning cached results can be significantly faster, as the database won't have to recompute the results to load your question.

You can set [caching invalidation policies](#cache-invalidation-policies) for [questions](#question-caching-policy), [dashboards](#dashboard-caching-policy), and [databases](#database-caching-policy).

## How caching works in Metabase

Let's say you set a caching policy for a particular question. You set a [duration](#duration-caching-policy) that says to invalidate the results after one hour.

When you view the question for the first time, Metabase will check for stored results. When it doesn't find any, it will run the question, return the results, and store (cache) the results. Those results will remain valid for the next hour (according to the hour-long duration policy you set).

> You run the question half an hour later, Metabase will return those stored results.

Reviewer: ```suggestion
If you run the question half an hour later, Metabase will return those stored results.
```

I realize what you have is good, but I figure this is a lil bit harder to misunderstand

> You run the question over an hour after that initial run, Metabase will notice that the stored results are older than the caching policy allows, so Metabase will delete the stored results, re-run the query, return the results, and store the results anew. These updated results will remain valid for the next hour.

Reviewer: ```suggestion
If you run the question over an hour after that initial run, Metabase will notice that the stored results are older than the caching policy allows. It will delete the stored results, run the query again, and return the results, caching them for the future. This cache will remain valid for the next hour.
```

For example, if your data only updates once a day, there's no point in querying the database more than once a day, as the data won't have changed. Returning cached results can be significantly faster, as the database won't have to recompute the results to load your question.

You can set [caching invalidation policies](#cache-invalidation-policies) for [questions](#question-caching-policy), [dashboards](#dashboard-caching-policy), and [databases](#database-caching-policy).

## Cache invalidation policies

> These policies determine how long cached results are valid.

Reviewer: In the recording I say that when the cache is invalidated, it is not deleted - no, I was wrong, the cache is deleted when it is invalidated. That's why I wondered if "clearing" or "deleting" the cache is friendlier than the comp-sci term "invalidation".

{% include plans-blockquote.html feature="Duration caching policy" %}

> Keep the cache for X number of hours. When someone runs a query, Metabase will first check if it's cached the results. If not, it runs the query and caches the results for as long as you set the duration.

Reviewer: Maybe:

`Keep the cache for a certain number of hours.`

Maybe this is just me but I trip a little over "it's cached" because I first think that's short for "it is"

Reviewer: Perhaps "caches the results for the specified duration"?

### Schedule caching policy

> 

Reviewer: <!--__GRAPHITE_HTML_TAG_START__--><i class='graphite__hidden'>[Re: line 37]<br/><br/></i>

<!--__GRAPHITE_HTML_TAG_END__-->Maybe "You can schedule the cache to invalidate"?

The bullet points work, I think, but a list with commas might also work well<!--__GRAPHITE_HTML_TAG_START__--><span class='graphite__hidden'><br/><br/>See this comment inline on <a href="https://app.graphite.dev/github/pr/metabase/metabase/45584?utm_source=unchanged-line-comment">Graphite</a>.</span>

<!--__GRAPHITE_HTML_TAG_END__-->

> 

Reviewer: <!--__GRAPHITE_HTML_TAG_START__--><i class='graphite__hidden'>[Re: line 51]<br/><br/></i>

<!--__GRAPHITE_HTML_TAG_END__-->Repeated sentence here<!--__GRAPHITE_HTML_TAG_START__--><span class='graphite__hidden'><br/><br/>See this comment inline on <a href="https://app.graphite.dev/github/pr/metabase/metabase/45584?utm_source=unchanged-line-comment">Graphite</a>.</span>

<!--__GRAPHITE_HTML_TAG_END__-->

4. Select the [caching invalidation policy](#cache-invalidation-policies).
5. Save your changes.

## How dashboard, question, database, and default caching policies interact

> If you have multiple caching policies set up, Metabase will use the first available policy, in this order:

Reviewer: In my head we either "set" or "assign" a caching policy. I'm less sure of "set up".

1. Dashboard
2. Question
3. Database
4. Default (site-wide)

A dashboard policy overrides a question policy, which overrides a database policy, which overrides a default policy.

> So if you give a question a caching policy, add that question to a dashboard, then set a caching policy for that dashboard, Metabase would use the dashboard's policy. But that's only when you run the dashboard. If you run the question by itself (outside of the dashboard), Metabase will use the question's policy. If the question appears in multiple dashboards, each with their own policy, the question will use whatever policy is relevant to the current dashboard.

Reviewer: I got a bit mixed up while reading this and wonder if there's an even simpler way to say it. Hard, though, because the logic just is fairly complicated.

Author: We're probably better omitting this explanation entirely, as it's harder to follow than the simple cache hierarchy.

4. Select the [caching invalidation policy](#cache-invalidation-policies).
5. Save your changes.

## How dashboard, question, database, and default caching policies interact

> If you have multiple caching policies set, Metabase will use the first available policy, in this order:

Reviewer: Just a passing idea

```suggestion
If multiple caching policies affect the same question, Metabase will use the first available policy, in this order:
```
